cy.selectFile docs

BlueWinds opened the pull request:
- Closes https://github.com/cypress-io/cypress-documentation/issues/4197

selectFile is a new command we'll be adding. This is ready for eyes on the PR, but should not be merged until the actual feature is ready to go, and it is still under development.

Tasks TODO:
- [x] Complete code work on feature.
- [x] Verify this PR matches the Cypress version we'll actually release attachFile in.

jennifer-shehane committed:
9.1.0 Release (#4224)

Co-authored-by: Emily Rohrbough <emilyrohrbough@users.noreply.github.com>

F3n67u committed:
Add `declare global` to type argument snippet (#4113)

khashaba committed:
Documentation - remove duplicate text in google authentication (#4230)

BlueWinds committed:
First draft of attachFile docs

BlueWinds committed:
Add migration guide for cypress-file-upload

jaffrepaul commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+---
```
Be sure to add the new page to https://github.com/cypress-io/cypress-documentation/blob/master/content/_data/sidebar.json so it gets a spot in the list & API TOC.

jennifer-shehane replied:
You'll also need to add the new page to `content/_data/en.json`

Does `attachFile` also follow Actionability logic? If so we'll want to add it as a file that follows the logic to the `content/guides/core-concepts/interacting-with-elements.md` page.

BlueWinds replied:
Added in all three places.

jaffrepaul commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+When clicking on `click` within the command log, the console outputs the
+following:
+
+<DocsImage src="/img/api/click/attach-file-in-console.png" alt="console.log for attachFile" ></DocsImage>
```
Did you want this image in the click folder? Will need either an updated path or to pop the png in `/attachfile`

BlueWinds replied:
Just a copy paste error, fixed.

jaffrepaul commented on content/guides/references/migration-guide.md:
```diff
@@ -2,6 +2,151 @@
+## Attaching Files
+
+Attaching files to input elements or dropping them over the page is available
+in Cypress 10.0. Read the [`cy.attachFile()` API docs](/api/commands/attachFile)
```
Camelcase will 404 here

api/commands/attachfile
```suggestion
in Cypress 10.0. Read the [`cy.attachFile()` API docs](/api/commands/attachFile)
```
```suggestion
in Cypress 10.0. Read the [`cy.attachFile()` API docs](/api/commands/attachFile)
```

BlueWinds replied:
Fixing.

Note that urls are actually case-insensitive - https://DOCS.cyPrEss.iO/api/cOmMaNdS/wItHiN works just fine. :D

jaffrepaul replied:
Ah ok. Likely because I was looking at this before the addition to the sidebar.json and assumed that was the issue. I'm going to write all my URLs in this fashion moving forward 😆

jennifer-shehane commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+title: attachFile
+---
+
+Attaches files to HTML5 input element, or simulated dragging files onto the browser.
```
```suggestion
Attaches a file or files to an HTML5 input element or simulates dragging a file or files into the browser.
```

jennifer-shehane commented on content/api/commands/attachfile.md:
````diff
@@ -0,0 +1,213 @@
+
+```javascript
+cy.attachFile('file.json') // Errors, cannot be chained off 'cy'
+cy.get('input[type=file]').attachFile('file contents') // Will attempt to find a file called 'file contents' on disk, probably not what you intended
````
```suggestion

// Will attempt to find a file called 'file contents' on disk, probably not what you intended
cy.get('input[type=file]').attachFile('file contents') 
```

jennifer-shehane commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+
+**<Icon name="angle-right"></Icon> file** **_(String, Array, Object or Buffer)_**
+
+Either a single file, or an array of them. A file can be:
+- A path to a file within the project root (the directory that contains the default `cypress.json` configuration file). Eg: `'path/to/file.json'`
+- `@alias` - An alias of any type, previously stored by `.as()`. Eg: `'@alias'`
+- A `Buffer()` containing binary data, such as that returned by `cy.readFile('file.json', { encoding: null })`. Eg: `Buffer.from('foo')`
+- An object with a non-null `contents` property, specifying details about the file. Eg: `{contents: '@alias', fileName: 'file.json'}`
```
Avoid pronouns, instead be more specific to avoid what 'them' is referring to. 

```suggestion
Either a single file, or an array of files. A file can be:

- A path to a file within the project root (the directory that contains the default Cypress configuration file). Eg: `'path/to/file.json'`
- `@alias` - An alias of any type, previously stored using `.as()`. Eg: `'@alias'`
- A `Buffer()` containing binary data, such as that returned by `cy.readFile('file.json', { encoding: null })`. Eg: `Buffer.from('foo')`
- An object with a non-null `contents` property, specifying details about the file. Eg: `{contents: '@alias', fileName: 'file.json'}`
```

jaffrepaul commented on content/api/commands/attachfile.md:
````diff
@@ -0,0 +1,213 @@
+cy.get('input[type=file]').attachFile('file.png')
+```
+
+If given a path, `cy.readFile()` will attach the file exactly as it exists on disk. This is the preferred way to work with files on disk, avoiding many encoding-related pitfalls.
````
```suggestion
If given a path, `.attachFile()` will attach the file exactly as it exists on disk. This is the preferred way to work with files on disk, avoiding many encoding-related pitfalls.
```

jennifer-shehane replied:
typo

```suggestion
If given a path, `.attachFile()` will attach the file exactly as it exists on disk. This is the preferred way to work with files on disk, avoiding many encoding-related pitfalls.
```

jaffrepaul commented on content/api/commands/attachfile.md:
````diff
@@ -0,0 +1,213 @@
+
+## Syntax
+
+```javascript
````
This could be confusing since the method must be chained. I thinks it's ok to just indicate each as `.attachFile(foo)`, like we do in the `.within()` page.

BlueWinds replied:
Good call, done.

jennifer-shehane commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+| -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| `contents`     | The contents of the file. This can be a string shorthand as described above, a `Buffer()` containing binary data or a non-Buffer object, which will be converted into a string with `JSON.stringify()` and utf8 encoded. |
+| `fileName`     | The name of the file. If `contents` is a path on disk, this defaults to the actual filename. In any other case, this defaults to an empty string.                                                                        |
+| `lastModified` | The file's last modified timestamp. This defaults to `Date.now()`.                                                                                                                                                       |
```
Might be good to clarify the time format expected (I guess saying Date.now() kind of does that). I can just see people passing in different datetime formats or something and not understanding why it's not working.

BlueWinds replied:
Done. Since I keep having to look up exactly how to get the timestamp from a js Date, I'm just adding a MDN link straight to the docs.

jennifer-shehane commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+| `animationDistanceThreshold` | [`animationDistanceThreshold`](/guides/references/configuration#Actionability) | The distance in pixels an element must exceed over time to be [considered animating](/guides/core-concepts/interacting-with-elements#Animations). |
+| `force`             | `false`                                                               | Forces the action, disables [waiting for actionability](#Assertions)                                                                                                |
+| `log`               | `true`                                                                | Displays the command in the [Command log](/guides/core-concepts/test-runner#Command-Log)                                                                            |
+| `timeout`           | [`defaultCommandTimeout`](/guides/references/configuration#Timeouts)  | Time to wait for `.find()` to resolve before [timing out](#Timeouts)                                                                                                |
```
typo

```suggestion
| `timeout`           | [`defaultCommandTimeout`](/guides/references/configuration#Timeouts)  | Time to wait for `.attachFile()` to resolve before [timing out](#Timeouts)                                                                                                |
```

jennifer-shehane commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+
+Depending on the action set in the `options` argument, `cy.attachFile()` can simulate two different user behaviors:
+
+- `input` (default)
+By default, `attachFile()` runs in 'input' mode, mimicking a user a selecting one or more files on an HTML5 input element. In this mode, the subject [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management) must be a single `input` element with `type="file"`, or a `label` element connected to one (either with its `for` attribute or by containing the input) ...
+- `drag-n-drop`
+Setting the action to `drag-n-drop` changes the behavior of the command to instead mimic a user dragging files from the operating system into the browser, and dropping them over the selected subject. In this mode, the subject can be any DOM element, or the `document` as a whole.
```
```suggestion
- `input` (default)
By default, `attachFile()` runs in 'input' mode, mimicking a user selecting one or more files on an HTML5 input element. In this mode, the subject [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management) must be a single `input` element with `type="file"`, or a `label` element connected to an input (either with its `for` attribute or by containing the input).
- `drag-n-drop`
Setting the action to `drag-n-drop` changes the behavior of the command to instead mimic a user dragging files from the operating system into the browser, and dropping them over the selected subject. In this mode, the subject can be any DOM element or the `document` as a whole.
```

jaffrepaul commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+
+| Version                                     | Changes                                                  |
+| ------------------------------------------- | -------------------------------------------------------- |
+| [10.0.0](/guides/references/changelog#10.0.0) | `cy.attachFile()` command added                        |
```
```suggestion
| [10.0.0](/guides/references/changelog#10.0.0) | `.attachFile()` command added                        |
```

jennifer-shehane commented on content/api/commands/attachfile.md:
````diff
@@ -0,0 +1,213 @@
+cy.get('input[type=file]').attachFile('file.json', { force: true })
+```
+
+In many cases in modern applications, the underlaying file input is hidden from view, and activated by a user clicking on a button. In these cases, you will need to tell Cypress to ignore its actionability checks and attach the file even though a user would not be able to directly activate the file input.
````
typo

```suggestion
In many cases in modern applications, the underlying file input is hidden from view, and activated by a user clicking on a button. In these cases, you will need to tell Cypress to ignore its actionability checks and attach the file even though a user would not be able to directly activate the file input.
```

jaffrepaul commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+
+### Action
+
+Depending on the action set in the `options` argument, `cy.attachFile()` can simulate two different user behaviors:
```
```suggestion
Depending on the action set in the `options` argument, `.attachFile()` can simulate two different user behaviors:
```

jaffrepaul commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+
+#### Default file existence assertion
+
+When passed a file path, `cy.attachFile()` asserts that the file exists and will fail if it does not exist. It will retry reading the file if it does not initially exist until the file exists or the command times out.
```
```suggestion
When passed a file path, `.attachFile()` asserts that the file exists and will fail if it does not exist. It will retry reading the file if it does not initially exist until the file exists or the command times out.
```

jennifer-shehane commented on content/api/commands/attachfile.md:
````diff
@@ -0,0 +1,213 @@
+cy.get('input[type=file]').attachFile('@myFixture')
+```
+
+Note the use of `null` encoding. By default, `cy.fixture()` `cy.readFile()` attempt to interpret files read from disk, which would result in a json file being decoded and re-encoded as a utf-8 string - the contents would be preserved, but formatting would not be, and the encoding might change. See [`cy.fixture`](/api/commands/fixture) or [`cy.readFile`](/api/commands/readfile) for more details on ...
````
```suggestion
Note the use of `null` encoding. By default, `cy.fixture()` and `cy.readFile()` attempt to interpret files read from disk, which would result in a JSON file being decoded and re-encoded as a utf-8 string - the contents would be preserved, but formatting would not be and the encoding might change. See [`cy.fixture`](/api/commands/fixture) or [`cy.readFile`](/api/commands/readfile) for more details on file encoding.
```

jaffrepaul commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+### Requirements [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Chains-of-Commands)
+
+<List><li>`.attachFile()` requires being chained off a command that yields DOM
+element(s). With the `input` action (default), it further requires a single `input` element with `type="file"`, or a `label` element attached to one.</li><li>If given a path, `cy.attachFile()` requires the file must exist.</li><li>If given an alias, `cy.attachFile()` requires that the subject of the alias must not be `null` or `undefined`</li></List>
```
```suggestion
element(s). With the `input` action (default), it further requires a single `input` element with `type="file"`, or a `label` element attached to one.</li><li>If given a path, `.attachFile()` requires the file must exist.</li><li>If given an alias, `.attachFile()` requires that the subject of the alias must not be `null` or `undefined`</li></List>
```
```suggestion
element(s). With the `input` action (default), it further requires a single `input` element with `type="file"`, or a `label` element attached to one.</li><li>If given a path, `.attachFile()` requires the file must exist.</li><li>If given an alias, `.attachFile()` requires that the subject of the alias must not be `null` or `undefined`</li></List>
```

jennifer-shehane commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+
+<DocsImage src="/img/api/attachfile/attach-file-during-test.png" alt="Command log for attachFile" ></DocsImage>
+
+When clicking on `click` within the command log, the console outputs the
```
typo

```suggestion
When clicking on `attachFile` within the command log, the console outputs the
```

jaffrepaul commented on content/guides/references/migration-guide.md:
```diff
@@ -2,6 +2,151 @@
+
+In the second argument:
+- `subjectType`: Rename to `action`.
+- `allowEmpty`: Remove. `cy.attachFile()` does not check the length of a file read from disk, only its existence.
```
```suggestion
- `allowEmpty`: Remove. `.attachFile()` does not check the length of a file read from disk, only its existence.
```

jennifer-shehane commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+
+| Version                                     | Changes                                                  |
+| ------------------------------------------- | -------------------------------------------------------- |
+| [10.0.0](/guides/references/changelog#10.0.0) | `cy.attachFile()` command added                        |
```
Does this have to go into 10.0.0? I feel like this can go in an earlier release. Usually we wait to add the History table until we have the release branch.

```suggestion
| [10.0.0](/guides/references/changelog#10.0.0) | `cy.attachFile()` command added                        |
```

BlueWinds replied:
So the reason I was thinking 10.0 is because of the popularity of the existing community solution, and the fact that adding a command that shares a name will throw an error if the plugin is still installed. While we did discuss 'not every fix is breaking' in standup today, and I agree it can sometimes be murky, but I think this one should be in a major version.

cypress-upload-file has [762 thousand](https://www.npmjs.com/package/cypress-file-upload) (!!!) weekly downloads - that's more a quarter as many downloads as Cypress itself. When we add in our own version of this command, every single one of those user's tests will break and they'll need to follow a migration guide.

It may theoretically be a new feature (minor release), but in practice it'll require a large portion of our users to update their code.

cowboy replied:
If this is going into 10.0.0, it would be nice to see it based off of the `unification` branch. Otherwise, carry on!

BlueWinds replied:
Switched to targeting the unification branch, thanks. I was looking for a 10.0-release branch matching the name of the cypress branch, or the 9.2-release branch naming we have in this repo and didn't see it.

jaffrepaul commented on content/guides/references/migration-guide.md:
```diff
@@ -2,6 +2,151 @@
+<Badge type="success">After</Badge> Working with file contents before attaching with
+Cypress 10.0. The `null` encoding introduced in Cypress 9.0 makes
+working with binary data simpler, and is the preferred encoding for use with
+`cy.attachFile()`.
```
```suggestion
`.attachFile()`.
```

jennifer-shehane commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,213 @@
+
+- [Guide: Variables and Aliases](/guides/core-concepts/variables-and-aliases)
+- [`.get()`](/api/commands/get)
+- [`.readFile()`](/api/commands/readfile)
```
cy.fixture

BlueWinds replied:
Added.

jennifer-shehane commented on content/guides/references/migration-guide.md:
```diff
@@ -2,6 +2,151 @@
+in Cypress 10.0. Read the [`cy.attachFile()` API docs](/api/commands/attachFile)
+for more information on how this works and how to use it.
+
+The
+[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
+plugin has been deprecated in favor of `attachFile()` built into Cypress.
+There's guidance below on how to migrate from the
+[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
+plugin to Cypress's built-in attachFile command.
```
```suggestion
The
[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
plugin has been deprecated in favor of `.attachFile()` built into Cypress.
There's guidance below on how to migrate from the
[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
plugin to Cypress's built-in attachFile command.
```

jennifer-shehane commented on content/guides/references/migration-guide.md:
```diff
@@ -2,6 +2,151 @@
+[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
+plugin to Cypress's built-in attachFile command.
+
+#### Quick guide
+In the first argument:
```
```suggestion
#### Quick guide

The argument signature is different for `.attachFile()` than the `cypress-file-upload` plugin. You can follow the steps below for each argument in order to migrate:

In the first argument:
```

jennifer-shehane reviewed (changes requested):
@BlueWinds This is great! The tone and level of explanation is exactly on point for our docs. I have some changes to request and we need to make sure the new doc appears in the sidebar so everyone can find it.

jaffrepaul reviewed (changes requested):
👋  @BlueWinds Exciting to get this in the mix! 

I'll give the content another run through but in the meantime I added a few housekeeping items that will help with the review.

Commit:
Apply suggestions from code review

Co-authored-by: Jennifer Shehane <jennifer@cypress.io>
Co-authored-by: Paul Jaffre <jaffrepaul@gmail.com>

BlueWinds committed:
Additional review changes

BlueWinds commented:
Thanks both of you for the proofreading - it's hard to check yourself for clarity and grammar sometimes. I've applied all the suggestions so far.

jennifer-shehane commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,217 @@
+Depending on the action set in the `options` argument, `.attachFile()` can simulate two different user behaviors:
+
+- `input` (default)
+By default, `attachFile()` runs in 'input' mode, mimicking a user selecting one or more files on an HTML5 input element. In this mode, the subject [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management) must be a single `input` element with `type="file"`, or a `label` element connected to an input (either with its `for` attribute or by containing the inp ...
```
This icon is not standard to use inline with sentences, we just link the word. 

```suggestion
By default, `attachFile()` runs in 'input' mode, mimicking a user selecting one or more files on an HTML5 input element. In this mode, the [subject](/guides/core-concepts/introduction-to-cypress#Subject-Management) must be a single `input` element with `type="file"`, or a `label` element connected to an input (either with its `for` attribute or by containing the input).
```

BlueWinds replied:
> A few other things I noticed while reading through the updated preview.
> 
> Some of the code blocks are showing scrollbars at the max size. Can you update the code blocks so you don't have to scroll to read? See preview: [deploy-preview-4241--cypress-docs.netlify.app/api/commands/attachfile#Arguments](https://deploy-preview-4241--cypress-docs.netlify.app/api/commands/attachfile#Arguments)

Went through and fixed all those.

BlueWinds replied:
Ah, whoop, last one wasn't supposed to be a child comment. Also fixed this.

jennifer-shehane commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,217 @@
+
+Depending on the action set in the `options` argument, `.attachFile()` can simulate two different user behaviors:
+
+- `input` (default)
```
The formatting of this looks strange when rendered. I think these should be H4s, or I think you need to indent the content within the li. See preview: https://deploy-preview-4241--cypress-docs.netlify.app/api/commands/attachfile#Arguments

jennifer-shehane commented on content/api/commands/attachfile.md:
````diff
@@ -0,0 +1,217 @@
+### From an API response
+
+```javascript
+cy.request('http://localhost:8888/users/827').get('body').as('responseBody')
````
This isn't correct. `cy.get()` cannot be chained off of another command. It's a parent command.

BlueWinds replied:
You're right, the example is incorrect. Oddly though, it doesn't throw an error:
```
cy.request('http://example.com').get('body').then(a => { console.log(a) })
// a is now the <body> element of the AUT
```
Maybe something to consider fixing with `.get()` at some point, making it stricter.

Will fix the example.

jennifer-shehane replied:
Yah, we never got around to enforced chaining. That would be a major breaking change.

jennifer-shehane reviewed (changes requested):
A few other things I noticed while reading through the updated preview.

Some of the code blocks are showing scrollbars at the max size. Can you update the code blocks so you don't have to scroll to read? See preview: https://deploy-preview-4241--cypress-docs.netlify.app/api/commands/attachfile#Arguments

<img width="771" alt="Screen Shot 2021-12-02 at 10 17 29 AM" src="https://user-images.githubusercontent.com/1271364/144461312-f90b8c67-250b-4f69-8c91-2b411bad99ab.png">

BlueWinds committed:
More review comments

BlueWinds committed:
Remove trailing newline

BlueWinds committed:
Add additional timeout information

Commit:
Merge branch 'unification' into issue-4197-attachFile

BlueWinds committed:
Fix merge

jennifer-shehane commented:
@jaffrepaul Can we get a rereview from you to approve this?

jaffrepaul commented on content/api/commands/attachfile.md:
````diff
@@ -0,0 +1,231 @@
+
+**<Icon name="check-circle" color="green"></Icon> Correct Usage**
+
+```javascript
````
We can get a little more of this section in view by removing the extra couple of spaces in front of this fenced in block.

```suggestion
```javascript
cy.get('input[type=file]').attachFile('file.json')
cy.get('input[type=file]').attachFile(['file.json', 'file2.json'])
cy.get('input[type=file]').attachFile({
  contents: Buffer.from('foo'),
  fileName: 'file.json',
  lastModified: Date.now(),
})
```

jaffrepaul replied:
This one may have gotten wonky via GH suggestions with the multiple sets of ```. The idea is just to close the extra white space.

BlueWinds replied:
Having the spaces looks nicer in the markdown, but markdown isn't the actual target format. :)

Done.

jaffrepaul commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,231 @@
+| ------------------- | --------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `action`            | `'input'`                                                             | Switches attachFile between modes. Valid values are `input` and `drag-n-drop`. See [Actions](#Actions) below for more details.                                      |
+| `animationDistanceThreshold` | [`animationDistanceThreshold`](/guides/references/configuration#Actionability) | The distance in pixels an element must exceed over time to be [considered animating](/guides/core-concepts/interacting-with-elements#Animations). |
+| `force`             | `false`                                                               | Forces the action, disables [waiting for actionability](#Assertions)                                                                                                |
```
```suggestion
| `force`             | `false`                                                               | Forces the action, disables [waiting for actionability](#Assertions).                                                                                                |
```

jaffrepaul commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,231 @@
+| `action`            | `'input'`                                                             | Switches attachFile between modes. Valid values are `input` and `drag-n-drop`. See [Actions](#Actions) below for more details.                                      |
+| `animationDistanceThreshold` | [`animationDistanceThreshold`](/guides/references/configuration#Actionability) | The distance in pixels an element must exceed over time to be [considered animating](/guides/core-concepts/interacting-with-elements#Animations). |
+| `force`             | `false`                                                               | Forces the action, disables [waiting for actionability](#Assertions)                                                                                                |
+| `log`               | `true`                                                                | Displays the command in the [Command log](/guides/core-concepts/test-runner#Command-Log)                                                                            |
```
```suggestion
| `log`               | `true`                                                                | Displays the command in the [Command log](/guides/core-concepts/test-runner#Command-Log).                                                                            |
```

jaffrepaul commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,231 @@
+| `animationDistanceThreshold` | [`animationDistanceThreshold`](/guides/references/configuration#Actionability) | The distance in pixels an element must exceed over time to be [considered animating](/guides/core-concepts/interacting-with-elements#Animations). |
+| `force`             | `false`                                                               | Forces the action, disables [waiting for actionability](#Assertions)                                                                                                |
+| `log`               | `true`                                                                | Displays the command in the [Command log](/guides/core-concepts/test-runner#Command-Log)                                                                            |
+| `timeout`           | [`defaultCommandTimeout`](/guides/references/configuration#Timeouts)  | Time to wait for `.attachFile()` to resolve before [timing out](#Timeouts)                                                                                                |
```
```suggestion
| `timeout`           | [`defaultCommandTimeout`](/guides/references/configuration#Timeouts)  | Time to wait for `.attachFile()` to resolve before [timing out](#Timeouts).                                                                                                |
```

jaffrepaul commented on content/api/commands/attachfile.md:
````diff
@@ -0,0 +1,231 @@
+])
+```
+
+This will fail unless the file input has the `multiple` property.
````
This is probably worth throwing in a lil alert to make sure it doesn't get lost within in the content.

```suggestion
<Alert type="warning">

This will fail unless the file input has the `multiple` property.

</Alert>
```

jaffrepaul commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,231 @@
+### Requirements [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Chains-of-Commands)
+
+<List><li>`.attachFile()` requires being chained off a command that yields DOM
+element(s). With the `input` action (default), it further requires a single `input` element with `type="file"`, or a `label` element attached to one.</li><li>If given a path, `.attachFile()` requires the file must exist.</li><li>If given an alias, `.attachFile()` requires that the subject of the alias must not be `null` or `undefined`</li></List>
```
```suggestion
element(s). With the `input` action (default), it further requires a single `input` element with `type="file"`, or a `label` element attached to one.</li><li>If given a path, `.attachFile()` requires the file must exist.</li><li>If given an alias, `.attachFile()` requires that the subject of the alias must not be `null` or `undefined`.</li></List>
```

jaffrepaul commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,231 @@
+### Assertions [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Assertions)
+
+<List><li>`.attachFile()` will automatically wait for the element to reach an
+[actionable state](/guides/core-concepts/interacting-with-elements)</li></List>
```
```suggestion
[actionable state](/guides/core-concepts/interacting-with-elements).</li></List>
```

BlueWinds replied:
I sure do love my missing periods!

jaffrepaul commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,154 @@ on('<event>', (arg1, arg2) => {
+
+#### Quick guide
+
+The argument signature is different for `.attachFile()` than the `cypress-file-upload` plugin. You can follow the steps below for each argument in order to migrate:
```
This is technically correct. I've only learned this recently because I have kids but they hate when people correct them on it 🤣 

```suggestion
The argument signature is different for `.attachFile()` from the `cypress-file-upload` plugin. You can follow the steps below for each argument in order to migrate:
```

BlueWinds replied:
It's not different for attachFiles() *from* cypress-file-upload though, that would mean something different!

I'll tweak the working a bit.

jaffrepaul commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,154 @@ on('<event>', (arg1, arg2) => {
+The argument signature is different for `.attachFile()` than the `cypress-file-upload` plugin. You can follow the steps below for each argument in order to migrate:
+
+In the first argument:
+- `filePath`: Prefix value with `fixtures/'. Rename to `contents`.
```
Was formatting awkwardly. I believe the second line needed to be removed, right? In favor of the next line? 

```suggestion
- `filePath`: Prefix value with `fixtures/`.
```

BlueWinds replied:
I added the missing tick.

In cypress-flie-upload, you could only use one at a time - providing both together was an error. So there are really two cases: Rename filePath to contents, *or* rename fileContent to contents. I'm open to better ways to explain that in the migration guide, but I think it would be clear enough when a user sees their code

```
{
  filePath: 'foo.json',
  encoding: 'utf-8',
}
```

that it turns into `{contents: 'fixture/foo.json'}` and that when they have

```
{
  fileContent: Blob.atob([91, 10]),
  mimeType: 'text/plain',
}
```

turns into `{content: Buffer.from([91, 10])}`.

I can't provide specific guidance on the Blob -> Buffer change because it's very application specific.

jaffrepaul replied:
Got it. thx!

jaffrepaul commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,154 @@ on('<event>', (arg1, arg2) => {
+
+<Badge type="success">After</Badge> Attaching a fixture from disk with
+Cypress 10.0. Cypress follows paths from the root of your test folder
+(as [`cy.readFile()`](/api/commands/readfile)).
```
Might be more clear (if only slightly)

```suggestion
(same as [`cy.readFile()`](/api/commands/readfile)).
```

jaffrepaul commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,154 @@ on('<event>', (arg1, arg2) => {
+
+<Badge type="success">After</Badge> Attaching a fixture from disk with
+Cypress 10.0. Cypress follows paths from the root of your test folder
+(as [`cy.readFile()`](/api/commands/readfile)).
```
```suggestion
(same as [`cy.readFile()`](/api/commands/readfile)).
```

jaffrepaul commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,154 @@ on('<event>', (arg1, arg2) => {
+
+<Badge type="success">After</Badge> Attaching a fixture from disk with
+Cypress 10.0. Cypress follows paths from the root of your test folder
+(as [`cy.readFile()`](/api/commands/readfile)).
```
```suggestion
(same as [`cy.readFile()`](/api/commands/readfile)).
```

jaffrepaul reviewed (changes requested):
@BlueWinds I left a handful of small suggestions. With that I think we're good to go. Great stuff!

BlueWinds committed:
Review changes

BlueWinds committed:
More changes

jaffrepaul reviewed (approved):
LGTM 🎉

cowboy commented on content/api/commands/attachfile.md:
````diff
@@ -0,0 +1,235 @@
+```javascript
+.attachFile(file)
+.attachFile(file, options)
+.attachFile([file], options)
+.attachFile([file], options)
````
I feel like the `.attachFile([file], options)` lines are unnecessary here

BlueWinds replied:
There definitely shouldn't be two of them, I'll fix that.

Even though it's not strictly speaking a different signature (still 1-2 args), it seemed worth highlighting, right at the top, that you can pass in an array of files. Perhaps `.attachFile([file, file2, ...], options)` would be clearer in that intention? Or do you not think it's worth having additional lines here at all?

cowboy replied:
I think we could do

```js
.attachFile([file, file2, ...], options)
```

or 

```js
.attachFile(fileArray, options)
```

but in either case, we probably also want to show the non-`options` variant, so maybe something like this?

```js
.attachFile(file)
.attachFile(file, options)
.attachFile(fileArray)
.attachFile(fileArray, options)
```

BlueWinds replied:
Yeah, that's why there was a duplicate line, accidentally had `options` in both. Pushed up changes.

cowboy commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,235 @@
+
+### Action
+
+Depending on the action set in the `options` argument, `.attachFile()` can simulate two different user behaviors:
```
```suggestion
Depending on the `action` set in the `options` argument, `.attachFile()` can simulate two different user behaviors:
```

cowboy commented on content/api/commands/attachfile.md:
````diff
@@ -0,0 +1,235 @@
+cy.get('input[type=file]').attachFile('file.png')
+```
+
+If given a path, `.attachFile()` will attach the file exactly as it exists on disk. This is the preferred way to work with files on disk, avoiding many encoding-related pitfalls.
````
Also reinforcing that files are loaded relative to the project root

```suggestion
If given a path, `.attachFile()` will attach the file, relative to the project root, exactly as it exists on disk. This is the preferred way to work with files on disk, avoiding many encoding-related pitfalls.
```

cowboy commented on content/api/commands/attachfile.md:
````diff
@@ -0,0 +1,235 @@
+### From a file on disk
+
+```javascript
+cy.get('input[type=file]').attachFile('file.json')
````
It might be helpful to reinforce that the file can have a path

```suggestion
cy.get('input[type=file]').attachFile('path/to/file.json')
```

BlueWinds replied:
Done.

cowboy commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,235 @@
+
+#### Default file existence assertion
+
+When passed a file path, `.attachFile()` asserts that the file exists and will fail if it does not exist. It will retry reading the file if it does not initially exist until the file exists or the command times out.
```
Does it also do this with an object with a `contents` property with a file path string?

BlueWinds replied:
Yep. Clarified slightly.

cowboy commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,235 @@
+#### The element must first reach actionability
+
+`.attachFile()` is an "action command" that follows all the rules
+[defined here](/guides/core-concepts/interacting-with-elements).
```
```suggestion
of [Actionability](/guides/core-concepts/interacting-with-elements#Actionability).
```

BlueWinds replied:
I copied this directly from another command, and it seems to be the common pattern in our docs.

![image](https://user-images.githubusercontent.com/3003404/146038924-2b69cf3b-a210-445b-a2aa-105b6ec5b90e.png)

Do you want to change this everywhere?

cowboy replied:
I think we should change it everywhere, but that can also be done in a separate, follow-up PR instead of further complicating this one. If we do that, just leave it as-is. Your call!

BlueWinds replied:
It's a good change but a small one, so I went ahead an applied it to all the docs here.

cowboy commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,235 @@
+
+### Requirements [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Chains-of-Commands)
+
+<List><li>`.attachFile()` requires being chained off a command that yields DOM
```
Can these be markdown lists?

BlueWinds replied:
I am not exactly sure why they're not, but I was copying the pattern from other files in this directory - they universally use `<List>` in their Rules section. Without understanding Chesterton's fence, I didn't want to change it. I'm wondering if maybe some of our custom scripts rely on it?

![image](https://user-images.githubusercontent.com/3003404/146043293-ac22fc7d-7348-4df7-abce-fe33296defa2.png)

cowboy replied:
I'm not entirely sure what's up with `<List>`, but the `List` component is just this:

```vue
<template>
  <ul>
    <slot />
  </ul>
</template>
```

I loaded all the commit diffs containing `<List` into vscode with:

```sh
git log -G'<List' -p | code -
```

and then I searched for `(?<![</])\bList\b` to try to find any references to `List` that aren't preceded by `<` or `/` (to filter out `<List>` and `</List>` and found nothing of interest.

I think we're ok making it a markdown list.

BlueWinds committed:
Review changes

BlueWinds committed:
Change wording around actionability link

BlueWinds committed:
Fix migration guide to properly show fixture paths

BlueWinds committed:
Add contributors section to attachFile docs

BlueWinds committed:
Add contributors section to attachFile docs

BlueWinds committed:
Merge branch 'issue-4197-attachFile' of github.com:cypress-io/cypress-documentation into issue-4197-attachFile

BlueWinds commented on content/api/commands/attachfile.md:
```diff
@@ -0,0 +1,239 @@
+| --------------------------------------------- | -------------------------------------------------------- |
+| [10.0.0](/guides/references/changelog#10.0.0) | `.attachFile()` command added                            |
+
+### Community Recognition
```
@jennifer-shehane , @cowboy , @jaffrepaul - Calling this section out as it's newly added since the last round of reviews. Just cleaning up some loose ends.

We don't really have a standard for this, but I borrowed the format from https://github.com/cypress-io/cypress/blob/develop/npm/cypress-schematic/README.md#community-recognition.

jaffrepaul commented on content/api/commands/attachfile.md:
```diff
@@ -223,9 +223,13 @@ following:
+
+### Community Recognition
+
+The `.attachFile()` command draws heavy inspiration from the now-deprecated Cypress File Upload plugin. It was make possible by contributors to the [@abramenal/cypress-file-upload](https://github.com/abramenal/cypress-file-upload) repository.
```
Typo in there. _It was made possible..._

cowboy commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,155 @@ on('<event>', (arg1, arg2) => {
+
+The
+[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload)
+plugin has been deprecated in favor of `.attachFile()` built into Cypress.
```
We should explicitly state that the cypress-file-upload is incompatible with 10.0, and will no longer work.

BlueWinds replied:
Done.

cowboy commented on content/guides/references/migration-guide.md:
````diff
@@ -83,6 +83,155 @@ on('<event>', (arg1, arg2) => {
+
+
+```js
+cy.get('[data-cy="file-input"]').attachFile('cypress/fixtures/myfixture.json');
````
Given the note
> `filePath`: Prefix value with `fixtures/`. Rename to `contents`.

Should this example be:
```js
cy.get('[data-cy="file-input"]').attachFile('fixtures/myfixture.json');
```

BlueWinds replied:
This section is correct, I've updated the quick guide to match.

cowboy reviewed (changes requested):
A few other comments. Also, can you run prettier on your files? It's configured as a `lint:markdown` yarn script, but for some reason isn't being enforced. I'll have to look into that separately.

BlueWinds committed:
Replace <List> with markdown lists, run prettier

BlueWinds commented:
> A few other comments. Also, can you run prettier on your files? It's configured as a `lint:markdown` yarn script, but for some reason isn't being enforced. I'll have to look into that separately.

Probably because I never ran `yarn` on the repository, so husky never installed the precommit hooks on my system. Fixed that. But it should probably fail the tests, so that we can catch this thing sooner in the future (especially for outside contributors).

BlueWinds committed:
Rename attachFile -> selectFile; Move to 9.3 release

BlueWinds commented:
Renamed attachFile to selectFile, and moved the release from 10.0 to 9.3. Will follow up once merged with a separate PR to bring these into the 9.3 release branch.

jennifer-shehane commented on content/api/commands/selectfile.md:
```diff
@@ -0,0 +1,276 @@
+title: selectFile
+---
+
+Attaches a file or files to an HTML5 input element or simulates dragging a file
+or files into the browser.
```
```suggestion
Selects a file or files to be uploaded to an HTML5 input element or simulates dragging a file
or files into the browser.
```

BlueWinds replied:
Going for a slightly different edit here:

> Selects a file or files to attach to an HTML5 input element or simulates dragging a file or files into the browser.

I don't necessarily think we need to avoid the word attach, since it's still more descriptive of what's actually happening than 'upload' is.

jennifer-shehane replied:
It's replacing 'attach' with 'to be uploaded' not directly the verb 'upload' since it's not uploaded yet. I think upload is more descriptive. If you look at the docs we sighted, like mdn - it uses attach 2 times and upload 35 times for instance. When I search for this feature in google I always type 'upload files'. I don't think attach is more descriptive than upload. @cowboy thoughts?

jennifer-shehane commented on content/api/commands/selectfile.md:
````diff
@@ -0,0 +1,276 @@
+cy.get('input[type=file]').selectFile('path/to/file.png')
+```
+
+If given a path, `.selectFile()` will search for the file relative to the
+project root and attach the file exactly as it exists on disk. This is the
+preferred way to work with files on disk, avoiding many encoding-related
+pitfalls.
````
```suggestion
If given a path, `.selectFile()` will search for the file relative to the
project root and upload the file exactly as it exists on disk. This is the
preferred way to work with files on disk, avoiding many encoding-related
pitfalls.
```

BlueWinds replied:
Going to leave this one as-is, since as mentioned above, I think attach is more descriptive than upload in this context.

jennifer-shehane commented on content/api/commands/selectfile.md:
````diff
@@ -0,0 +1,276 @@
+cy.get('input[type=file]').selectFile('file.json', { force: true })
+```
+
+In many cases in modern applications, the underlying file input is hidden from
+view, and activated by a user clicking on a button. In these cases, you will
+need to tell Cypress to ignore its actionability checks and attach the file even
+though a user would not be able to directly activate the file input.
````
```suggestion
In many cases in modern applications, the underlying file input is hidden from
view, and activated by a user clicking on a button. In these cases, you will
need to tell Cypress to ignore its actionability checks and upload the file even
though a user would not be able to directly activate the file input.
```

BlueWinds replied:
Going to use 'select' rather than 'attach' here.

jennifer-shehane commented on content/api/commands/selectfile.md:
````diff
@@ -0,0 +1,276 @@
+cy.get('input[type=file]').selectFile('@myFile')
+```
+
+### Attaching multiple files
````
```suggestion
### Selecting multiple files
```

jennifer-shehane commented on content/api/commands/selectfile.md:
```diff
@@ -0,0 +1,276 @@
+
+</Alert>
+
+### Attaching a file with custom filename and lastModified
```
```suggestion
### Selecting a file with custom filename and lastModified
```

jennifer-shehane commented on content/api/commands/selectfile.md:
```diff
@@ -0,0 +1,276 @@
+
+## Command Log
+
+**_Attach file to input_**
```
```suggestion
**_Select file to upload to input_**
```

BlueWinds replied:
Applying slightly different version of this one

jennifer-shehane commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,170 @@ on('<event>', (arg1, arg2) => {
 
 CONTENT_TBD
 
+## Migrating to Cypress 9.3
+
+This guide details the changes and how to change your code to make full use of
+Cypress 9.3.
+[See the full changelog for 9.3](/guides/references/changelog#9-3-0).
+
+None of the changes detailed here are breaking.
```
I would get rid of this section and just make the main title 'Migrating to selectFile'. This is similar to the section on cy.route to cy.intercept migration which was in a point release.

BlueWinds replied:
Ah, that is much nicer. Changing now.

jennifer-shehane commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,170 @@ on('<event>', (arg1, arg2) => {
+
+None of the changes detailed here are breaking.
+
+### Attaching Files
+
+Attaching files to input elements or dropping them over the page is available in
+Cypress 9.3. Read the [`.selectFile()` API docs](/api/commands/selectfile) for
+more information on how this works and how to use it.
```
```suggestion
### Selecting Files to Upload

Selecting files to upload to input elements or dropping them over the page is available in
Cypress 9.3. Read the [`.selectFile()` API docs](/api/commands/selectfile) for
more information on how this works and how to use it.
```

BlueWinds replied:
Slightly different tweak since I'm reworking the introduction to match the other style.

jennifer-shehane commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,170 @@ on('<event>', (arg1, arg2) => {
+
+#### Quick guide
+
+The argument signature is different for Cypress' builtin `.selectFile()` command
+than `.attachFile` the `cypress-file-upload` plugin provided. You can follow the
+steps below for each argument in order to migrate:
```
```suggestion
The argument signature is different for Cypress' builtin `.selectFile()` command
than the `.attachFile` command the `cypress-file-upload` plugin provided. You can follow the
steps below for each argument in order to migrate:
```

jennifer-shehane commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,170 @@ on('<event>', (arg1, arg2) => {
+
+### Examples
+
+Below are several examples of migrating various commands from
+`cypress-file-upload` to the builtin `.attachFile()`.
```
```suggestion
Below are several examples of migrating various commands from
`cypress-file-upload` to the builtin `.selectFile()` command.
```

jennifer-shehane commented on content/guides/references/migration-guide.md:
````diff
@@ -83,6 +83,170 @@ on('<event>', (arg1, arg2) => {
+})
+```
+
+<Badge type="success">After</Badge> Attaching a fixture from disk with
````
```suggestion
<Badge type="success">After</Badge> Selecting a fixture from disk with
```

jennifer-shehane commented on content/guides/references/migration-guide.md:
````diff
@@ -83,6 +83,170 @@ on('<event>', (arg1, arg2) => {
+})
+```
+
+<Badge type="success">After</Badge> Attaching a fixture from disk with
````
```suggestion
<Badge type="success">After</Badge> Selecting a fixture from disk with
```

jennifer-shehane commented on content/guides/references/migration-guide.md:
````diff
@@ -83,6 +83,170 @@ on('<event>', (arg1, arg2) => {
+cy.get('[data-cy="file-input"]').attachFile('myfixture.json')
+```
+
+<Badge type="success">After</Badge> Attaching a fixture from disk with
````
```suggestion
<Badge type="success">After</Badge> Selecting a fixture from disk with
```

jennifer-shehane commented on content/guides/references/migration-guide.md:
````diff
@@ -83,6 +83,170 @@ on('<event>', (arg1, arg2) => {
+  })
+```
+
+<Badge type="success">After</Badge> Working with file contents before attaching
````
```suggestion
<Badge type="success">After</Badge> Working with file contents before selecting
```

BlueWinds replied:
Picked "using" rather than "selecting" here.

jennifer-shehane commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,170 @@ on('<event>', (arg1, arg2) => {
+
+#### Working with file contents
+
+<Badge type="danger">Before</Badge> Working with file contents before attaching
```
```suggestion
<Badge type="danger">Before</Badge> Working with file contents before selecting
```

BlueWinds replied:
Picked "using" rather than "selecting" here.

jennifer-shehane commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,170 @@ on('<event>', (arg1, arg2) => {
+<Badge type="success">After</Badge> Working with file contents before attaching
+with `.selectFile()`. The `null` encoding introduced in Cypress 9.0 makes
+working with binary data simpler, and is the preferred encoding for use with
+`.attachFile()`.
```
```suggestion
`.selectFile()`.
```

jennifer-shehane reviewed (changes requested):
Need to also rename the folder that the image assets are within to selectFile.

cowboy commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,170 @@ on('<event>', (arg1, arg2) => {
 
 CONTENT_TBD
 
+## Migrating to Cypress 9.3
```
Instead of having this content in the Migration Guide, which is really geared toward mandatory changes, can you add a section called "Upgrading from the cypress-file-upload plugin" to the `.selectFile()` API docs?

BlueWinds replied:
I prefer the solution we used with https://docs.cypress.io/guides/references/migration-guide#Migrating-cy-route-to-cy-intercept, where it's in the migration guide but not attached to a specific release. Moving it into the selectFile docs would clutter up our API docs with a guide that's not of interest to new users or old users once they've gone through it once. It also feels like it would be harder to find in there.

BlueWinds commented on content/guides/core-concepts/introduction-to-cypress.md:
````diff
@@ -530,17 +530,17 @@ cy.visit('/my/resource/path') // Nothing happens yet
 
 ```js
-cy.mount(<MyComponent />)     // Nothing happens yet
+cy.mount(<MyComponent />) // Nothing happens yet
````
Done.

Commit:
Apply suggestions from code review

Co-authored-by: Jennifer Shehane <jennifer@cypress.io>

BlueWinds committed:
Review changes

BlueWinds commented:
All review comments addressed - I think we're about ready to merge into unification, so that I can start on a PR for master (for when we're ready to release 9.3 in January)?

jennifer-shehane reviewed (dismissed):
The images file location needs to be updated to be within a selectfile folder. 

I still don't agree with the use of the word 'attach' as the first sentence of the selectFile doc and think it should be replaced with 'to be uploaded'. I left a comment.

BlueWinds committed:
Merge remote-tracking branch 'origin/unification' into issue-4197-attachFile

BlueWinds committed:
Final review comment, move images to proper folder

BlueWinds commented:
Removed the use of "attach" from the opening sentence but shortening it to remove the second verb.

```
Selects a file or files in an HTML5 input element or simulates dragging a file
or files into the browser.
```

Also moved the image files to the proper directory.

BlueWinds committed:
Remove link to example we're also removing

jennifer-shehane commented:
@BlueWinds I changed the base branch to 9.3.0-release branch. It seems there are some conflicts from the branch update. Can you resolve these?

BlueWinds commented:
> @BlueWinds I changed the base branch to 9.3.0-release branch. It seems there are some conflicts from the branch update. Can you resolve these?

Discussed in Slack, changing back to unification. Will make a second PR to add these changes to 9.3.0-release.

cowboy commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,158 @@ on('<event>', (arg1, arg2) => {
+Selecting files with input elements or dropping them over the page is available
+in Cypress 9.3. Read the [`.selectFile()` API docs](/api/commands/selectfile)
+for more information on how this works and how to use it. This guide details how
+to change your test code to migrate from
```
Can we say "migrate from the cypress-file-upload plugin" instead of "migrate from cypress-file-upload" here?

BlueWinds replied:
Done.

cowboy commented on content/api/commands/selectfile.md:
````diff
@@ -0,0 +1,276 @@
+  fileName: 'file.json',
+  lastModified: Date.now(),
+})
+```
````
Can we show one or more drag+drop examples here, eg.

```js
cy.get('.file-target').selectFile('file.json', { action: 'drag-drop' })
cy.document().selectFile('file.json', { action: 'drag-drop' })
```

BlueWinds replied:
Added.

cowboy commented on content/api/commands/selectfile.md:
```diff
@@ -0,0 +1,276 @@
+cy.get('input[type=file]').selectFile('file.json')
+cy.get('input[type=file]').selectFile(['file.json', 'file2.json'])
+cy.get('input[type=file]').selectFile({
+  contents: Buffer.from('foo'),
```
Can we change `foo` to `file contents` to better mirror the "incorrect usage" example below? And also change `file.json` to something like `file.txt` because either of the above are invalid JSON

BlueWinds replied:
Done and done.

cowboy commented on content/api/commands/selectfile.md:
```diff
@@ -0,0 +1,276 @@
+
+| Option                       | Default                                                                        | Description                                                                                                                                       |
+| ---------------------------- | ------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `action`                     | `'select'`                                                                     | Switches selectFile between modes. Valid values are `select` and `drag-drop`. See [Actions](#Actions) below for more details.                     |
```
Can you either format selectFile as `.selectFile()` here or remove "selectFile between"?

BlueWinds replied:
Removed extra words.

cowboy commented on content/api/commands/selectfile.md:
```diff
@@ -0,0 +1,276 @@
+
+#### `select` (default)
+
+By default, `selectFile()` runs in 'select' mode, mimicking a user selecting one
```
```suggestion
By default, `.selectFile()` runs in 'select' mode, mimicking a user selecting one
```

BlueWinds replied:
Applied as part of other commits.

cowboy commented on content/api/commands/selectfile.md:
````diff
@@ -0,0 +1,276 @@
+```javascript
+cy.get('input[type=file]').selectFile({
+  contents: 'path/to/file.json',
+  fileName: 'users.json',
````
Maybe this can be more clear like so:

```suggestion
  fileName: 'custom-name.json',
```

BlueWinds replied:
Done.

cowboy commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,158 @@ on('<event>', (arg1, arg2) => {
+than the `.attachFile` command the `cypress-file-upload` plugin provided. You
+can follow the steps below for each argument in order to migrate:
+
+In the first argument:
+
+- `filePath`: Prefix value with `cypress/fixtures/`. Rename to `contents`.
+- `fileContent`: Use `Buffer.from()` rather than `Cypress.Blob` methods. Rename
+  to `contents`.
+- `encoding`: Remove. No longer needed due to improved binary file handling in
+  Cypress 9.0.
+- `mimeType`: Remove. No longer needed due to improved binary file handling in
+  Cypress 9.0.
+
+In the second argument:
+
+- `subjectType`: Rename to `action`. Move from `drag-n-drop` to `drag-drop` or
+  from `input` to `select`.
+- `allowEmpty`: Remove. `.selectFile()` does not check the length of a file read
+  from disk, only its existence.
```
Could you change this section to something along these lines? I want to be as clear as possible with the language:

When the first argument is a string filepath (or array of string filepaths):

- Prefix the existing filepath(s) with `cypress/fixtures/`.

When the first argument is an object:

- `filePath`: Rename this property to `contents`. Prefix the existing value with `cypress/fixtures/`.
- `fileContent`: Rename this property to `contents`. Use `Buffer.from()` rather than `Cypress.Blob` methods. 
- `encoding`: Remove this property. No longer needed due to improved binary file handling in Cypress 9.0.
- `mimeType`: Remove this property. No longer needed due to improved binary file handling in Cypress 9.0.

The second argument:

- `subjectType`: Rename this property to `action`. Change the existing value from `drag-n-drop` to `drag-drop`, or from `input` to `select`.
- `allowEmpty`: Remove this property. The `.selectFile()` command does not check the length of a file read from disk, only its existence.

BlueWinds replied:
Good suggestions, applied.

cowboy commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,158 @@ on('<event>', (arg1, arg2) => {
+In the first argument:
+
+- `filePath`: Prefix value with `cypress/fixtures/`. Rename to `contents`.
+- `fileContent`: Use `Buffer.from()` rather than `Cypress.Blob` methods. Rename
```
Do we need to explain to the user where `Buffer` is coming from here? Are they going to be confused?

BlueWinds replied:
I've added a link to the nodejs docs for Buffer. Questions like "why can I use Buffer in my cypress code, I thought it executed in the browser" are beyond the scope of a migration guide.

cowboy commented on content/api/commands/selectfile.md:
```diff
@@ -0,0 +1,276 @@
+  default Cypress configuration file). Eg: `'path/to/file.json'`
+- `@alias` - An alias of any type, previously stored using `.as()`. Eg:
+  `'@alias'`
+- A `Buffer()` containing binary data, such as that returned by
```
Same question here as in the migration guide WRT `Buffer`. Are people going to understand what a `Buffer` is? Do we need to link to something here or explain this below?

BlueWinds replied:
Added the same link to the node docs here.

cowboy replied:
Does cypress provide `Buffer` for people to use, or is it something they will need to import in their spec file? It may be useful to show that in an example on this page, because it doesn't seem obvious (like, I personally do not know what I would need to do to use `Buffer` like in these examples, and would have to google it)

BlueWinds replied:
Buffer is a global glass, same as Array or Object. I don't think there's anything more that needs to be done to make this explicit - the examples are complete. There's no import statement needed. If they copy paste the examples exactly as written, they will just work.

We don't explicitly call out that `cy` is provided as a global either - we just use it in our examples and assume people will understand and try it. :shrug:

cowboy replied:
Ok, cool. As long as it works!

cowboy replied:
I just ran a few tests with this branch. When using component testing, projects that use Webpack will polyfill `Buffer`, so your examples work as-is. However, projects that use Vite do not polyfill `Buffer`, so we end up seeing a `Buffer is not defined` error, eg.

![image](https://user-images.githubusercontent.com/54051/148593663-58f5db22-90df-4206-9a8d-d83064697b65.png)

We can work around this issue by polyfilling `Buffer` with the [buffer](https://www.npmjs.com/package/buffer) npm module (per this issue https://github.com/vitejs/vite/discussions/2785) and then doing something like this at the top of any spec files that use `Buffer`:

```js
import { Buffer } from 'buffer'
```

or globally in the `supportFile`:

```js
import { Buffer } from 'buffer'
globalThis.Buffer = Buffer
```

We should add a note to this effect somewhere!

cowboy replied:
@BlueWinds I'm happy to take a stab at writing this, just LMK and I'll make a commit.

BlueWinds replied:
Huh, today I learned that we ship with a Vite based dev server! https://docs.cypress.io/guides/component-testing/framework-configuration#Vite-Based-Projects-Vue-React

I'm getting set up with a vite project now - ideally we'd solve this in our code rather than our documentation, so that users have a consistent experience. Now I'm wondering if the readFile option to use Buffers works there at all, or if it's been broken since release and no one has noticed.

BlueWinds replied:
The link you provided suggests we should be able to do `globalThis.Buffer = Buffer` just once, and it should work everywhere (including in user's spec file).

cowboy replied:
If you need examples, also see:

- https://github.com/cypress-io/cypress-component-testing-examples/tree/main/vite-react
- https://github.com/cypress-io/cypress-component-testing-examples/tree/main/vite-vue

BlueWinds committed:
Review feedback

cowboy commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,164 @@ on('<event>', (arg1, arg2) => {
+
+When the first argument is a file path:
+
+- Prefix the path with `cypress/fixtures`
```
```suggestion
- Prefix the path with `cypress/fixtures/`.
```

BlueWinds replied:
Applied.

cowboy commented on content/guides/references/migration-guide.md:
```diff
@@ -83,6 +83,164 @@ on('<event>', (arg1, arg2) => {
+[`cypress-file-upload`](https://github.com/abramenal/cypress-file-upload) plugin
+to `.selectFile()`.
+
+#### Quick guide
```
I feel like all `####` headers in this PR can be changed to `###`, and the `Examples` section header can possibly go away. This would make the right-side nav look more consistent with that of other migration guide sections. Thoughts?

BlueWinds replied:
I'm changing quick guide to ###, but I like Examples as a ### because then it shows up in the table of contents, but each individual example doesn't. Since only 2nd and 3rd level headers are rendered:

![image](https://user-images.githubusercontent.com/3003404/148590245-6c98c27a-c50d-4715-aab1-9d798dbe15da.png)

A TOC that looks like

## Migrating from `cypress-file-upload` to [`.selectFile()`](/api/commands/selectfile)
### Quick Guide
### Examples

looks pretty clean while this one is overly verbose for the right-hand column TOC:

## Migrating from `cypress-file-upload` to [`.selectFile()`](/api/commands/selectfile)
### Quick Guide
### Read and attach a fixture
### Using drag-n-drop
### Overriding the file name
### Working with file contents

BlueWinds committed:
More tweaks

BlueWinds committed:
Add Cypress.Buffer docs

cowboy commented:
@BlueWinds is there a reason this PR hasn't been based off of the `9.3.0-release` branch? We're not going to merge `unification` until 10.0 is released.

BlueWinds commented:
> @BlueWinds is there a reason this PR hasn't been based off of the `9.3.0-release` branch? We're not going to merge `unification` until 10.0 is released.

The plan is to merge these changes into unification, then make a second PR to pull the commits into 9.3.0. This will be easier than attempting to rebase the branch onto 9.3.0.

Originally it was targeted at unification because we were going to release this alongside 10.0.

cowboy commented on content/api/commands/selectfile.md:
```diff
@@ -0,0 +1,281 @@
+  default Cypress configuration file). Eg: `'path/to/file.json'`
+- `@alias` - An alias of any type, previously stored using `.as()`. Eg:
+  `'@alias'`
+- A [`Cypress.Buffer()`](/api/utilities/buffer) containing binary data, such as
```
I don't think we want to show `Cypress.Buffer` being called as a function. How about this?

```suggestion
- A [`Cypress.Buffer`](/api/utilities/buffer) instance containing binary data, such as
```

cowboy commented on content/api/utilities/buffer.md:
```diff
@@ -0,0 +1,67 @@
+
+Use `Cypress.Buffer` to manipulate binary data, similar to the
+[nodejs Buffer class](https://nodejs.org/api/buffer.html). Instances are
+returned by `cy.readFile()` and `cy.fixture()`, and it is used with
```
Can we link the 3 cypress commands in this paragraph to their respective pages?

cowboy commented on content/api/utilities/buffer.md:
```diff
@@ -0,0 +1,67 @@
+
+## Examples
+
+### Image Fixture
```
The "Image Fixture" header seems a bit out of place, since it only seems to apply to the first example, but not the second.

BlueWinds committed:
Add Buffer to sidebar, links to it in fixture and readfile

cowboy commented on content/api/utilities/buffer.md:
````diff
@@ -0,0 +1,67 @@
+
+```javascript
+// Programmatically upload the logo
+cy.readFile('images/logo.png', null).then((file) => {
+  Cypress.Buffer.isBuffer(file) // true
+})
````
I'm concerned that if we show people `Cypress.Buffer.isBuffer(file)` like this that they might think it's going to implicitly assert like many cypress commands do. In this case, `file` will always be a `Cypress.Buffer` instance, but I'm concerned that someone might see this example, and then expect `Cypress.Buffer.isBuffer(nonBufferThing)` to fail the test like `cy.get('#non-existent-id')` would.

```suggestion
cy.readFile('images/logo.png', null).then((file) => {
  expect(Cypress.Buffer.isBuffer(file)).to.be.true
})
```

cowboy commented on content/api/utilities/buffer.md:
```diff
@@ -0,0 +1,67 @@
+const files = []
+
+for (var i = 1; i <= 10; i++) {
+  files.push(Buffer.from(`Contents of file #${i}`))
```
```suggestion
  files.push(Cypress.Buffer.from(`Contents of file #${i}`))
```

cowboy commented on content/api/commands/readfile.md:
```diff
@@ -46,8 +46,8 @@ supported:
-Using `null` explicitly will return the file as a `Buffer`, regardless of file
-extension.
+Using `null` explicitly will return the file as a
+[`Cypress.Buffer()`](/api/utilities/buffer), regardless of file extension.
```
```suggestion
[`Cypress.Buffer`](/api/utilities/buffer) instance, regardless of file extension.
```

cowboy commented on content/api/commands/fixture.md:
```diff
@@ -57,8 +57,8 @@ supported:
-Using `null` explicitly will return the fixture as a `Buffer`, regardless of
-file extension.
+Using `null` explicitly will return the fixture as a
+[`Cypress.Buffer()`](/api/utilities/buffer), regardless of file extension.
```
```suggestion
[`Cypress.Buffer`](/api/utilities/buffer) instance, regardless of file extension.
```

cowboy commented on content/api/commands/fixture.md:
```diff
@@ -248,7 +249,8 @@ Cypress automatically determines the encoding for the following file types:
 specified in the second argument of `cy.fixture()`. You can specify `null` as
-the encoding in order to read the file as a `Buffer` instead.
+the encoding in order to read the file as a
+[`Cypress.Buffer()`](/api/utilities/buffer) instead.
```
```suggestion
[`Cypress.Buffer`](/api/utilities/buffer) instance instead.
```

cowboy commented on content/api/commands/fixture.md:
```diff
@@ -144,6 +144,7 @@ cy.fixture('images/logo.png', null).then((logo) => {
   // logo will be read as a buffer
   // and should look something like this:
   // Buffer([0, 0, ...])
+  Cypress.Buffer.isBuffer(logo) // true
```
```suggestion
  expect(Cypress.Buffer.isBuffer(logo)).to.be.true
```

cowboy commented on content/api/commands/selectfile.md:
```diff
@@ -0,0 +1,281 @@
+
+| Option         | Description                                                                                                                                                                                                                                        |
+| -------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `contents`     | The contents of the file. This can be a string shorthand as described above, a `Cypress.Buffer()` containing binary data or a non-Buffer object, which will be converted into a string with `JSON.stringify()` and `utf8` encoded.                 |
```
```suggestion
| `contents`     | The contents of the file. This can be a string shorthand as described above, a `Cypress.Buffer` instance containing binary data or a non-Buffer object, which will be converted into a string with `JSON.stringify()` and `utf8` encoded.                 |
```

cowboy reviewed (changes requested):
Overall, looks great. Just a few minor changes and then it's good to go!

cowboy commented:
@BlueWinds Oops - I just noticed that the two screenshots are out of date and show `attachFile` and not `selectFile`.

Commit:
Apply suggestions from code review

Co-authored-by: Ben Alman <cowboy@rj3.net>

cowboy committed:
Resolve a few outstanding issues

cowboy committed:
Update selectFile screenshots

cowboy committed:
Fix linting issues

cowboy commented:
Closing in favor of https://github.com/cypress-io/cypress-documentation/pull/4310
